File: user-service/src/main/java/rs/raf/user_service/domain/mapper/RoleMapper.java
package rs.raf.user_service.domain.mapper;

import rs.raf.user_service.domain.dto.RoleDto;
import rs.raf.user_service.domain.entity.Role;

import java.util.stream.Collectors;

public class RoleMapper {

    public static RoleDto toDto(Role role) {
        if (role == null) return null;
        RoleDto dto = new RoleDto();
        dto.setId(role.getId());
        dto.setName(role.getName());
        if (role.getPermissions() != null) {
            dto.setPermissions(role.getPermissions().stream()
                    .map(PermissionMapper::toDTO)
                    .collect(Collectors.toSet()));
        }
        return dto;
    }
}

File: user-service-kotlin/src/main/kotlin/pack/userservicekotlin/domain/mapper/RoleMapper.kt
package pack.userservicekotlin.domain.mapper

import pack.userservicekotlin.domain.dto.role.RoleResponseDto
import pack.userservicekotlin.domain.entities.Role

fun Role?.toDto(): RoleResponseDto? {
    if (this == null) return null

    return RoleResponseDto(
        id = id,
        name = name,
        permissions = permissions.mapNotNull { it.toDto() }.toSet(),
    )
}
